web/packages/teleterm/src/services/config/getKeyboardShortcutSchema.ts

```diff
@@ -0,0 +1,210 @@
+    code: z.ZodIssueCode.custom,
+    message: `Duplicate modifier found.`,
+  };
+}
+
+export function missingModifierIssue(keyCode: string): z.IssueData {
+  return {
+    code: z.ZodIssueCode.custom,
+    message: `"${keyCode}" requires a modifier.`,
+  };
+}
+
+export function getKeyboardShortcutSchema(platform: Platform) {
+  const allowedModifiers = getSupportedModifiers(platform);
+
+  return z
+    .string()
+    .transform(s => s.trim().split(/\s?\+\s?/))
+    .transform(putModifiersFirst(allowedModifiers))
+    .superRefine(validateKeyCodeAndModifiers(allowedModifiers))
+    .transform(adjustCasing)
+    .transform(s => s.join('+'));
+}
+
+function putModifiersFirst(
+  allowedModifiers: string[]
+): (tokens: string[]) => string[] {
+  return tokens =>
+    tokens.sort((a, b) => {
+      if (allowedModifiers.indexOf(a) === -1) {
+        return 1;
+      }
+      if (allowedModifiers.indexOf(b) === -1) {
+        return -1;
+      }
+      return allowedModifiers.indexOf(a) - allowedModifiers.indexOf(b);
+    });
+}
+
+/** Currently, works only for single characters.
+ * An accelerator string is compared to a string generated from event properties
+ * (so it should be case-insensitive too).
+ * But what is more important, this string is used also for tooltips.
+ * If we allow "CTRL+SHIFT+PAGEUP" then we cannot display it as "Ctrl+Shift+PageUp".
+ */
+function adjustCasing(tokens: string[]): string[] {
+  return tokens.map(token => {
+    if (token.length === 1) {
+      return token.toUpperCase();
+    }
+    return token;
+  });
+}
+
+function validateKeyCodeAndModifiers(
+  allowedModifiers: string[]
+): (tokens: string[], ctx: z.RefinementCtx) => void {
+  return (tokens, ctx) => {
+    // empty accelerator disables the shortcut
+    if (tokens.join('') === '') {
+      return z.NEVER;
+    }
+
+    const [keyCode, ...modifiers] = [...tokens].reverse();
+
+    const keyCodeUppercase =
+      keyCode.length === 1 // letters
+        ? keyCode.toUpperCase()
+        : keyCode;
+    if (!ALLOWED_KEY_CODES.includes(keyCodeUppercase)) {
+      ctx.addIssue(invalidKeyCodeIssue(keyCode));
+      return z.NEVER;
+    }
+
+    if (modifiers.length !== new Set(modifiers).size) {
+      ctx.addIssue(duplicateModifierIssue());
+      return z.NEVER;
+    }
+
+    const invalidModifiers = modifiers.filter(
+      modifier => !allowedModifiers.includes(modifier)
+    );
+    if (invalidModifiers.length) {
+      ctx.addIssue(invalidModifierIssue(invalidModifiers));
+      return z.NEVER;
+    }
+
+    if (!FUNCTION_KEYS.includes(keyCode) && !modifiers.length) {
+      ctx.addIssue(missingModifierIssue(keyCode));
+    }
+  };
+}
+
+/** Returns allowed modifiers for a given platform in the correct order. */
+function getSupportedModifiers(platform: Platform): string[] {
+  switch (platform) {
+    case 'win32':
+    case 'linux':
+      return ['Ctrl', 'Alt', 'Shift'];
+    case 'darwin':
+      return ['Cmd', 'Ctrl', 'Option', 'Shift'];
```

Comment:
It feels weird to me now that there's Cmd and Ctrl but then you have Shift (which doesn't have a natural short name) and Option (which could be Opt or Alt). Ideally we'd support both full and short names but let's leave it like this for now.

Reply:
Yeah, I felt the same, and we changed it only to display the same shortcut in an error message for both platforms. 
Maybe we could tweak the error message and go back to `Command` and `Control`? Then all modifiers would have a consistent form.

Reply:
What do you mean by tweaking the error message? Sorry my brain barely functions atm.

Reply:
I think the error message is fine but yeah I guess let's go back to full spellings for now.

Reply:
Done. 
I also changed the order of modifiers for Mac (to follow [platform guidelines](https://developer.apple.com/design/human-interface-guidelines/inputs/keyboards/#custom-keyboard-shortcuts)). I thought it would only affect `useKeyboardShortcutsFormatters` but the actual sorting is done in `getKeyboardShortcutSchema` so I had to change it there (and in `keybaordShortcutsService`). But I think that's fine, we have a consistent order in the entire app.